juju/juju
# Have the SetCharmURL apiserver call wait for the cache.

This change introduces a primitive into the cache.Model where you can wait for a field to be updated. I agonised for a while about the best method to have a generic approach without adding significant overhead to places that don't warrent it. 

In the end I settled on specific published events inside the model hub for changes that can be listened to. It is expected that only a small subset of fields are actually interesting to wait on.

The apiserver facade now waits on the cache to be updated.

## QA steps

* run `stress -check.f SetCharmURL` in `apiserver/facades/agent/uniter`
* comment out the line in `uniter.go` where the facade call has `waitForControllerCharmURL`, and try again, mine failed on the first try.

## Bug reference

_there probably is that one about generations feature and deploying things_
